broker/src/main/java/io/camunda/zeebe/broker/bootstrap/AbstractBrokerStartupStep.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.broker.bootstrap;
+
+import static io.camunda.zeebe.util.sched.future.CompletableActorFuture.completedExceptionally;
+
+import io.camunda.zeebe.util.sched.ConcurrencyControl;
+import io.camunda.zeebe.util.sched.future.ActorFuture;
+import io.camunda.zeebe.util.startup.StartupStep;
+import java.util.function.BiConsumer;
+
+abstract class AbstractBrokerStartupStep implements StartupStep<BrokerStartupContext> {
+
+  @Override
+  public final ActorFuture<BrokerStartupContext> startup(
+      final BrokerStartupContext brokerStartupContext) {
+    return createFutureAndRun(
+        brokerStartupContext,
+        (concurrencyControl, future) ->
+            startupInternal(brokerStartupContext, concurrencyControl, future));
+  }
+
+  @Override
+  public final ActorFuture<BrokerStartupContext> shutdown(
+      final BrokerStartupContext brokerShutdownContext) {
+    return createFutureAndRun(
+        brokerShutdownContext,
+        (concurrencyControl, future) ->
+            shutdownInternal(brokerShutdownContext, concurrencyControl, future));
+  }
+
+  abstract void startupInternal(
+      final BrokerStartupContext brokerStartupContext,
+      final ConcurrencyControl concurrencyControl,
+      final ActorFuture<BrokerStartupContext> startupFuture);
+
+  abstract void shutdownInternal(
+      final BrokerStartupContext brokerShutdownContext,
+      final ConcurrencyControl concurrencyControl,
+      final ActorFuture<BrokerStartupContext> shutdownFuture);
+
+  /**
+   * Helper function that tries to create a future and call a runnable. If an exception is thrown
+   * while creating the future, then this exception is forwarded to a dummy future
+   */
+  final ActorFuture<BrokerStartupContext> createFutureAndRun(
+      final BrokerStartupContext brokerStartupContext,
+      final BiConsumer<ConcurrencyControl, ActorFuture<BrokerStartupContext>> runnable) {
+    try {
+      final var concurrencyControl = brokerStartupContext.getConcurrencyControl();
+      final ActorFuture<BrokerStartupContext> future = concurrencyControl.createFuture();
+
+      forwardExceptions(() -> runnable.accept(concurrencyControl, future), future);
+      return future;
+    } catch (final Throwable t) {
+      return completedExceptionally(t);
+    }
+  }
+  /**
+   * helper function that forwards exceptions thrown by a synchronous block of code to a future
+   * object
+   */
+  final <V> void forwardExceptions(final Runnable r, final ActorFuture<V> future) {
```

Comment:
:wrench: This is interesting, and I'd like to discuss this a bit more, but I'm also OK with accepting this for now. I don't think we should catch all `Throwable` however, as there is a small class which will mean the whole JVM will fail to do anything. See for example Vavr's [Try](https://www.javadoc.io/doc/io.vavr/vavr/0.9.2/io/vavr/control/Try.html) documentation which touches on fatal errors, and the discussion surrounding it [here](https://github.com/vavr-io/vavr/issues/2380) (which later links to Scala's [NonFatal](https://scala-lang.org/files/archive/api/2.13.x/scala/util/control/NonFatal$.html)). 

:thought_balloon: Which really makes me wish we had something like the `Try` or `Result` monad to wrap these kind of calls easily and chain them, but that's definitely out of scope. Let's say you had something like `Result<V>`, being a `Either<Throwable, V>`, or `Try`, being a `Either<Throwable, Void>`, then you could write the above as:

```java
  @Override
  public final ActorFuture<BrokerStartupContext> startup(
      final BrokerStartupContext brokerStartupContext) {
    final var concurrencyControl = brokerStartupContext.getConcurrencyControl();
    final ActorFuture<BrokerStartupContext> future = concurrencyControl.createFuture();

    Try.of(() -> startupInternal(brokerStartupContext, concurrencyControl, future))
        .orElseRun(future::completeExceptionally)

    return future;
  }
```